src/main.c

```diff
@@ -1157,6 +1163,63 @@ static yoml_t *load_config(const char *fn)
     return yoml;
 }
 
+static int filter_config_node(h2o_configurator_context_t *ctx, yoml_t **node)
+{
+    size_t i, j;
+
+    if ((*node)->type != YOML_TYPE_MAPPING)
+        return -1;
+
+    i = (*node)->data.mapping.size;
+    do {
+        --i;
+        if ((*node)->data.mapping.elements[i].key->type == YOML_TYPE_SCALAR &&
+            strcmp((*node)->data.mapping.elements[i].key->data.scalar, "include") == 0) {
+
+            yoml_mapping_element_t src = (*node)->data.mapping.elements[i];
+            if (src.value->type != YOML_TYPE_SCALAR)
+                return -1;
+            yoml_t *included = load_config(src.value->data.scalar);
+            if (included == NULL)
+                return -1;
+            if (included->type != YOML_TYPE_MAPPING) {
+                h2o_configurator_errprintf(NULL, included, "node must be a MAPPING");
+                return -1;
+            }
+
+            /* include recursively */
+            if (filter_config_node(ctx, &included) != 0)
+                return -1;
+
+            if (included->data.mapping.size > 1) {
+                /* expand elements preserving the values */
+                *node = realloc(*node, offsetof(yoml_t, data.mapping.elements) +
+                                ((*node)->data.mapping.size - 1 + included->data.mapping.size) * sizeof((*node)->data.mapping.elements[0]));
+                memmove((*node)->data.mapping.elements + i + included->data.mapping.size, (*node)->data.mapping.elements + i + 1,
+                        ((*node)->data.mapping.size - i - 1) * sizeof((*node)->data.mapping.elements[0]));
+                (*node)->data.mapping.size += included->data.mapping.size - 1;
+            }
+
+            for (j = 0; j != included->data.mapping.size; ++j) {
+                yoml_t *key = included->data.mapping.elements[j].key;
+                yoml_t *value = included->data.mapping.elements[j].value;
+
+                (*node)->data.mapping.elements[i + j].key = key;
+                ++key->_refcnt;
+                (*node)->data.mapping.elements[i + j].value = value;
+                ++value->_refcnt;
```

Comment:
Maybe copying as a struct (i.e. `(*node)->data.mapping.elements[i + j] = include->data.mapping.elements[j];`) and then adjusting the refcounts would simplify the code.